crates/agglayer-grpc-api/src/certificate_submission_service/mod.rs

```diff
@@ -37,6 +41,24 @@ where
         &self,
         request: tonic::Request<SubmitCertificateRequest>,
     ) -> Result<tonic::Response<SubmitCertificateResponse>, tonic::Status> {
+        // Retrieve extra signature from query metadata if any
+        let extra_signature: Option<Signature> = {
+            let metadata = request.metadata();
+            if let Some(value) = metadata.get(GRPC_METADATA_NAME_EXTRA_CERTIFICATE_SIGNATURE) {
+                let sig_str = value.to_str().map_err(|e| {
+                    tonic::Status::invalid_argument(format!("invalid signature encoding: {e}"))
+                })?;
+
+                let sig = Signature::from_str(sig_str).map_err(|e| {
+                    tonic::Status::invalid_argument(format!("invalid hex in signature: {e}"))
+                })?;
+
+                Some(sig)
+            } else {
+                None
+            }
```

Comment:
You can also write it this way. It really depends from the taste, feel free to ignore this suggestion.

```suggestion
                    request
                .metadata()
                .get(GRPC_METADATA_NAME_EXTRA_CERTIFICATE_SIGNATURE)
                .map(|value| {
                    value
                        .to_str()
                        .map_err(|e| {
                            tonic::Status::invalid_argument(format!(
                                "invalid signature encoding: {e}"
                            ))
                        })
                        .and_then(|sig_str| {
                            Signature::from_str(sig_str).map_err(|e| {
                                tonic::Status::invalid_argument(format!(
                                    "invalid hex in signature: {e}"
                                ))
                            })
                        })
                })
                .transpose()?
```